tests/pytorch/test_onnx_export.py

```diff
@@ -174,7 +196,8 @@ def create_ort_input_dict(session, inps):
             for loc in mismatched_ids[:nb_vals]:
                 ref = te_output[loc]
                 print(f"{onnx_output[loc]} -- {te_output[loc]} err={abs_err[loc]} > {atol + rtol * abs(ref)}")
-            raise ValueError(f"Output validation of {fname} failed with {len(mismatched_ids)} errors")
+            if len(mismatched_ids) > allow_cnt_errors:
+                raise ValueError(f"Output validation of {fname} failed with {len(mismatched_ids)} errors")
```

Comment:
If we wanted to be extra careful, we could check and make sure the errors are still within FP8 machine epsilon, something like `atol_fp8=0.15`.

Reply:
They are not within an fp8 range necessarily. I ran various experiments, running 1000 loops at a time. Depending on the exact thresholds that I ran I saw 1-6 cases where there were 1-3 outliers that have error values within some small range (I don't remember the values). There are 64x32 (2K) coefficients in the output, so in the 1-6 failing cases the error rate is 0.05%-0.15%.
So I pondered how to interpret these errors? What is their meaning within the context of these tests?
It means that for a very small set of random inputs TE and ORT compute Layer Norm outputs for which `abs(ort - te) > (atol + rtol * abs(te)`.

- Does that mean that the ONNX graph computes something different from what TE computes? -- No
- Does it mean that the ONNX graph and TE compute the same result -- Yes, because the errors can only be explained by small implementation differences.
- Let's imagine that the errors are "large" (keeping the same number of errors) - should we fail the test? I don't think so because this failure is still the result of TE vs ORT implementation differences.
- What value should `allow_cnt_errors` have? How do we choose it? I set a threshold and ran some 1000s test and collected statistics. I increased the threshold until I saw a maximum of 3 errors in each test running over 1000 tests, with ~99.95% of the test passing with 0 errors.

So I don't think that adding another hyperparameter to the test will make it more robust or more explainable.

Reply:
While ONNX and TE might have some implementation differences, they should be trying to approximate the same math operation. If the relative error is significantly larger than the expected numerical accuracy, that's a sign that at least one of the implementations is drifting far from the mathematical ground truth and should be a valid test failure. I can imagine bugs that might only show up in one or two entries, like if we skip the first entry.